service/frontend/health_check.go

```diff
@@ -91,15 +94,29 @@ func (h *healthCheckerImpl) Check(ctx context.Context) (enumsspb.HealthState, er
 	}
 
 	var failedHostCount float64
+	var startingHostCount float64
 	for i := 0; i < len(hosts); i++ {
 		healthState := <-receiveCh
-		if healthState == enumsspb.HEALTH_STATE_NOT_SERVING {
+		switch healthState {
+		case enumsspb.HEALTH_STATE_NOT_SERVING, enumsspb.HEALTH_STATE_UNSPECIFIED:
 			failedHostCount++
+		case enumsspb.HEALTH_STATE_STARTING:
+			startingHostCount++
+		case enumsspb.HEALTH_STATE_SERVING:
+			// Do nothing.
 		}
 	}
 	close(receiveCh)
 
-	if (failedHostCount / float64(len(hosts))) > h.hostFailurePercentage() {
+	startingHostCountPercentage := startingHostCount / float64(len(hosts))
+	if startingHostCountPercentage > h.hostStartingPercentage() {
+		h.logger.Warn("health check exceeded host starting percentage threshold", tag.NewFloat64("host starting percentage threshold", h.hostStartingPercentage()))
+		return enumsspb.HEALTH_STATE_STARTING, nil
+	}
+
+	failedHostCountPercentage := failedHostCount / float64(len(hosts))
+	if failedHostCountPercentage+startingHostCountPercentage > h.hostFailurePercentage() {
```

Comment:
I'm trying to think if it makes sense to include "starting" in with "failed" here... I think so but not 100% clear. Did you have a reason or example in mind when deciding that?

Reply:
Makes sense to me